Update distribution pipeline

psomhorst opened the pull request:
This PR started out small, but I added some improvements to several parts of the distribution pipeline.  

- To ensure `eitprocessing` keeps working on newer versions of Python, I added the test and build action for Python 3.11-3.13. 
- The image containing the data ([ghcr.io/eit-alive/eittestdata:latest](https://ghcr.io/eit-alive/eittestdata)) only has Python 3.10 with pip installed. ~Instead of editing that image, I edited the GitHub action to mount the data from the image, and run the test and build on `python:3.x-bookworm` instead. This makes it easier to adapt to new Python versions and replace the data image in the future, if necessary.~ See comment below for new description.
- The coveralls action was not working. I removed it from the test action and created a new action. Coveralls does not report well when run on pull requests (and they don't really make sense), so I restricted it to running on pushes to `main` and `develop`. 
- I updated the linting action to install the ruff version from pyproject.toml. I already forgot to update the version in the action once, not planning on doing it again.
- I updated the classifiers in the pyproject.toml file to better represent the (current status of) the project. 
- I added a py.typed file to the root of the package. This enables linters to use the source code for type checking when installing eitprocessing from package. (See https://peps.python.org/pep-0561/#packaging-type-information and https://mypy.readthedocs.io/en/stable/installed_packages.html#creating-pep-561-compatible-packages.)

DaniBodor commented on .github/workflows/build.yml:
```diff
@@ -16,67 +16,34 @@ on:
       matrix:
         os: ["ubuntu-latest"]
-        python-version: ["3.10"]
+        python-version: ["3.10", "3.11", "3.12", "3.13"]
```
For environmental reasons, I am not a big fan of running the CI on so many python versions (see [blog](https://blog.esciencecenter.nl/reduce-reuse-recycle-save-the-planet-one-github-action-at-a-time-4ab602255c3f)), given that at the end of the day, it's unlikely that it fails in only one.
While it's nice to offer the flexibility, I don't see much harm to instead state that we officially support any one python version but that it likely works on newer versions as well.

psomhorst replied:
Good suggestion! I agree that this is not very environmentally friendly. There are a few alternatives that reduce the number of runs while still ensuring it'll run on newer versions of Python. 

- Only run on the earliest and latest supported Python release. Is it safe to assume it will run on all intermittent versions?
- Stop official support for 3.10 and 3.11. Both only receive security fixes and should not be the base for new projects. Aggressive removal of supported versions means existing projects can only use new features if they update (assuming something actually breaks) or wing it (if their workflow still works). 
- Remove tests for all versions from the test_and_build action, but add it to the Draft Github Release action. This makes sure we don't release anything that breaks, but does not test every PR while it is being developed.

DaniBodor replied:
> Remove tests for all versions from the test_and_build action, but add it to the Draft Github Release action. This makes sure we don't release anything that breaks, but does not test every PR while it is being developed.

I really like this option, actually. Most breaks will be picked up in all versions anyway, and the few that are version specific can be picked up like this. The only thing to make sure of is to give a clear "error message" when it fails, because that could otherwise be very confusing.

DaniBodor commented:
This looks fine to me at first glance, but I am not very fluent at docker, so asked @wbaccinelli to review it instead. Also, things will change again soon in #303

psomhorst commented:
> This looks fine to me at first glance, but I am not very fluent at docker, so asked @wbaccinelli to review it instead. Also, things will change again soon in #303

I wasn't aware of #303. Maybe it's best to do that in this PR, instead of changing it twice? I think we can already remove the container requirement by extracting the data files from the container (https://github.com/marketplace/actions/docker-extract). This can then later be replaced, e.g. by downloading from Zenodo (although downloading from Zenodo seems to be quite slow). 

I'll look into this tomorrow.

psomhorst commented:
Since yesterday, I have changed a number of things:

- I renamed 'Build' to 'Test and build' to more aptly describe the action.
- I de-containerized the test and build action. It still uses the old container, but extracts the data from it for testing. The actual testing and building is done outside of a container (technically still inside one, but not in a container within a container). 
- I removed python 3.11-3.13 from the normal test and build action again. This reduces resource usage.
- I added a test job to the GitHub release action. This tests the package for python 3.10-3.13. This ensures we don't release anything that breaks the package, but is only run once per release instead of multiple times for every PR.
- I cleaned up the linting action, by splitting installing ruff, the linting and the formatting
- I updated the artifacts action to v4 (v3 will be deprecated soon)
- I updated all checkout actions to v4

The actions that run when a PR is updated (Test and build, Linting) work. 

The actions that run with a GitHub release can't be tested without creating a release. Since the tests have to succeed before the release is created, I dare to test them by making a release. If something is not right, the release will just fail.

psomhorst commented:
This PR closes #303.

psomhorst commented:
I now also updated the Coveralls action to de-containerize it.

DaniBodor commented on .github/workflows/coverage.yml:
a large proportion of this workflow is duplicated from the build workflow, which is not ideal because 1) if we want to change anything, it needs to be changed in multiple files, and 2) the env is being built and rebuilt multiple times.

This can be mitigated by creating a re-used workflow that builds the env and downloads the data, which can then be called multiple times to perform the different checks. See [here](https://docs.github.com/en/actions/sharing-automations/reusing-workflows).

DaniBodor commented on .github/workflows/lint.yml:
```diff
@@ -36,8 +36,11 @@ jobs:
-          ruff check || (echo "Please ensure you have a matching version of ruff (`ruff -V`) installed locally." && (exit 1))
-          ruff format --check || (echo "Please ensure you have a matching version of ruff (`ruff -V`) installed locally." && (exit 1))
+          python3 -m pip install pyproject-deplister
+          pyproject-deplister --extra dev --path pyproject.toml | grep ruff | sed 's/ //g' | xargs -I{} python3 -m pip install "{}"
```
nice! I will re-use this elsewhere too!

DaniBodor commented on .github/workflows/build.yml:
```diff
@@ -1,4 +1,4 @@
-name: Build package
+name: Test and build package
```
I personally prefer shorter names (both for the file and the job), as it keeps the checks list below inside the PRs tidier. Especially on smaller screens, the second half can drop off screen if names become too long.

DaniBodor commented on pyproject.toml:
This file is used for static type checking, right? We are not doing that in this package as of yet. 
I was looking into adding a static type checker (mypy) earlier this week, but I got too many errors and decided it wasn't enough of a priority to spend time resolving them.
Without that, does this file serve any purpose?

wbaccinelli commented on .github/workflows/build.yml:
```diff
@@ -16,67 +16,38 @@ on:
-            pytest --cov --cov-append --cov-report xml --cov-report term --cov-report html ;
-            coveralls --service=github ;
-            python${{ matrix.python-version }} -m build'
+      - name: Extract data from image
+        uses: shrink/actions-docker-extract@v3
+        with: 
+          image: ghcr.io/eit-alive/eittestdata:latest
+          destination: $EIT_PROCESSING_TEST_DATA
+          path: /eitprocessing/.
```
Same for coverage and test release

wbaccinelli reviewed (approved):
It looks good to me overall. The docker image will not be used in the near future, but we can leave it like this for now
